rama-net/src/client/pool.rs

```diff
@@ -0,0 +1,618 @@
+    pub fn take(mut self) -> C {
+        let conn = self.pooled_conn.take().expect("only None after drop").conn;
+        conn
+    }
+}
+
+impl<C, ConnId> Deref for LeasedConnection<C, ConnId> {
+    type Target = C;
+
+    fn deref(&self) -> &Self::Target {
+        &self
+            .pooled_conn
+            .as_ref()
+            .expect("only None after drop")
+            .conn
+    }
+}
+
+impl<C, ConnId> DerefMut for LeasedConnection<C, ConnId> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self
+            .pooled_conn
+            .as_mut()
+            .expect("only None after drop")
+            .conn
+    }
+}
+
+impl<C, ConnId> AsRef<C> for LeasedConnection<C, ConnId> {
+    fn as_ref(&self) -> &C {
+        self
+    }
+}
+
+impl<C, ConnId> AsMut<C> for LeasedConnection<C, ConnId> {
+    fn as_mut(&mut self) -> &mut C {
+        self
+    }
+}
+
+impl<C, ConnId> Drop for LeasedConnection<C, ConnId> {
+    fn drop(&mut self) {
+        if let (Some(pool), Some(pooled_conn)) = (self.pool.upgrade(), self.pooled_conn.take()) {
+            pool.return_pooled_conn(pooled_conn);
+        }
+    }
+}
+
+// We want to be able to use LeasedConnection as a transparent wrapper around our connection.
+// To achieve that we conditially implement all traits that are used by our Connectors
+
+impl<C: Socket, ConnId: Sync + Send + 'static> Socket for LeasedConnection<C, ConnId> {
+    fn local_addr(&self) -> std::io::Result<SocketAddr> {
+        self.as_ref().local_addr()
+    }
+
+    fn peer_addr(&self) -> std::io::Result<SocketAddr> {
+        self.as_ref().peer_addr()
+    }
+}
+
+impl<C: AsyncWrite + Unpin, ConnId: Unpin> AsyncWrite for LeasedConnection<C, ConnId> {
+    fn poll_write(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> std::task::Poll<Result<usize, std::io::Error>> {
+        // let x = self.deref_mut().deref_mut();
+        Pin::new(self.deref_mut().as_mut()).poll_write(cx, buf)
+    }
+
+    fn poll_flush(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), std::io::Error>> {
+        Pin::new(self.deref_mut().as_mut()).poll_flush(cx)
+    }
+
+    fn poll_shutdown(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), std::io::Error>> {
+        Pin::new(self.deref_mut().as_mut()).poll_shutdown(cx)
+    }
+}
+
+impl<C: AsyncRead + Unpin, R: Unpin> AsyncRead for LeasedConnection<C, R> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> std::task::Poll<std::io::Result<()>> {
+        Pin::new(self.deref_mut().as_mut()).poll_read(cx, buf)
+    }
+}
+
+impl<State, Request, C: Service<State, Request>, ConnId: Send + Sync + 'static>
```

Comment:
move these trait bounds to a `where` clause, it becomes unredable otherwise (IMHO) if you make the `impl<...>` part this long. For a simple (single) trait bound it's fine, otherwise better to a where clause, like is the case here.